p4-fusion/labels_conversion.cc

```diff
@@ -0,0 +1,212 @@
+	if (pos == std::string::npos)
+	{
+		throw std::invalid_argument("failed to parse commit message, well known section : change =  not found");
+	}
+	// The changelist number starts after 11 characters, the length of ": change = ".
+	const size_t clStart = pos + 11;
+	const size_t clEnd = message.find(']', clStart);
+	if (clEnd == std::string::npos)
+	{
+		throw std::invalid_argument("failed to parse commit message, well closing ] not found");
+	}
+
+	std::string cl(message, clStart, clEnd - clStart);
+
+	return cl;
+}
+
+int updateTags(P4API* p4, const std::string& depotPath, git_repository* repo)
+{
+	PRINT("Updatings tags")
+
+	// Load labels
+	PRINT("Requesting labels from the Perforce server")
+	LabelsResult labelsRes = p4->Labels();
+	if (labelsRes.HasError())
+	{
+		ERR("Failed to retrieve labels for mapping: " << labelsRes.PrintError())
+		return 1;
+	}
+	const std::list<std::string>& labels = labelsRes.GetLabels();
+	SUCCESS("Received " << labels.size() << " labels from the Perforce server")
+
+	std::unordered_map<std::string, std::unordered_map<std::string, LabelResult>*> revToLabel;
+
+	for (auto& label : labelsRes.GetLabels())
+	{
+		LabelResult labelRes = p4->Label(label);
+		if (labelRes.HasError())
+		{
+			ERR("Failed to retrieve label details: " << labelRes.PrintError());
+			continue;
+		}
+		if (!labelRes.revision.starts_with("@"))
+		{
+			continue;
+		}
+		labelRes.revision.erase(labelRes.revision.begin());
+		if (labelRes.views.empty())
+		{
+			if (!revToLabel.contains(labelRes.revision))
+			{
+				auto* newMap = new std::unordered_map<std::string, LabelResult>;
+				revToLabel.insert({ labelRes.revision, newMap });
+			}
+			auto res = revToLabel.at(labelRes.revision);
+			res->insert({ sanitizeLabelName(labelRes.label), labelRes });
+		}
+		else
+		{
+			for (auto& view : labelRes.views)
+			{
+				if (depotPath.starts_with(trimSuffix(view, "...")))
+				{
+					if (!revToLabel.contains(labelRes.revision))
+					{
+						auto* newMap = new std::unordered_map<std::string, LabelResult>;
+						revToLabel.insert({ labelRes.revision, newMap });
+					}
+					auto res = revToLabel.at(labelRes.revision);
+					res->insert({ sanitizeLabelName(labelRes.label), labelRes });
+				}
+			}
+		}
+	}
+
+	git_reference_iterator* refIter;
+	checkGit2Error(git_reference_iterator_glob_new(&refIter, repo, "refs/tags/*"));
+	git_reference* ref;
+	while (git_reference_next(&ref, refIter) >= 0)
+	{
+		std::string labelName = trimPrefix(git_reference_name(ref), "refs/tags/");
+
+		git_commit* commit;
+		checkGit2Error(git_commit_lookup(&commit, repo, git_reference_target(ref)));
+		if (std::string cl = getChangelistFromCommit(commit); revToLabel.contains(cl) && revToLabel.at(cl)->contains(labelName))
+		{
+			revToLabel.at(cl)->erase(labelName);
+			if (revToLabel.at(cl)->empty())
+			{
+				delete revToLabel.at(cl);
+				revToLabel.erase(cl);
+			}
+		}
+		else
+		{
+			PRINT("Tag has moved or no longer exists, deleting: " << labelName)
```

Comment:
I think the label changing/disappearing would cause this, not the tag, right?

```suggestion
			PRINT("Label has moved or no longer exists, deleting: " << labelName)
```